cmd/snap/cmd_run.go

```diff
@@ -493,30 +483,115 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
+// 	- Snap was removed.
+func isSnapRefreshConflictDetected(oldInfo *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
+	// XXX: An alternative for this complex check is to check if hint file exists.
+	// It didn't exist before and exists now hence a race condition is detected.
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+		// Skip check
+		return false
+	}
+
+	if !osutil.FileExists(filepath.Join(dirs.SnapMountDir, oldInfo.InstanceName(), "current")) {
+		// Current symlink does not exist, snap is either removed or being refreshed.
+		logger.Debugf("possible ongoing refresh conflict detected; current symlink does not exist")
+		return true
+	}
+
+	currentInfo, err := snap.ReadCurrentInfo(oldInfo.InstanceName())
+	if err != nil {
+		logger.Debugf("could not read snap info: %v", err)
+		return true
+	}
+	if currentInfo.SnapRevision() != oldInfo.SnapRevision() {
+		// Snap was refreshed while we did not hold the hint lock.
+		logger.Debugf("possibly snap was refreshed while snap-run was not holding the hint lock")
+		logger.Debugf("mismatch detected between current revision (%s) and initial revision (%s)", currentInfo.SnapRevision(), oldInfo.SnapRevision())
+		return true
+	}
+
+	return false
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
-	if err != nil {
-		return err
-	}
 
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
-	}
+	var retryCnt int
+	for {
+		if retryCnt > 1 {
+			// This should never happen, but it is better to fail instead
+			// of retrying forever.
+			return fmt.Errorf("internal error: race condition detected, snap-run can only retry once")
+		}
 
-	if !app.IsService() {
-		// TODO: use signal.NotifyContext as context when snap-run flow is finalized
-		if err := maybeWaitWhileInhibited(context.Background(), snapName); err != nil {
+		info, app, hintFlock, err := maybeWaitWhileInhibited(context.Background(), snapName, appName)
+		if err != nil {
 			return err
 		}
-	}
 
-	return x.runSnapConfine(info, app.SecurityTag(), snapApp, "", args)
+		var errSnapRefreshConflict = fmt.Errorf("snap refresh conflict detected")
+		closeFlockOrRetry := func() error {
+			// This needs to run inside the transient cgroup created for a snap
+			// such that any pending refresh of the snap will get blocked after
+			// we release the lock.
+			if hintFlock != nil {
+				// It is okay to release the lock here (beforeExec) because snapd unless forced
+				// will not inhibit the snap and do a refresh anymore because it detects app
+				// processes are running for via the established transient cgroup cgroup.
+				hintFlock.Close()
```

Comment:
Can the flock be set O_CLOEXEC so that this part becomes unnecessary and is instead guaranteed by the kernel?

Reply:
O_CLOEXEC is already being set, but the snap run could do a (fork and exec + waiting) in multiple cases so we might hold the lock indefinitely (locking up snapd writing):
https://github.com/snapcore/snapd/blob/953fad1e3cafca9545befc65dfb6c8327631300a/cmd/snap/cmd_run.go#L1254-L1273

Reply:
This is already O_CLOEXEC so this code is IMO not doing anything.

Reply:
Ah, I see. This is relevant indeed. Can you add a comment why we cannot rely O_CLOEXEC in the code please.

Reply:
Great point, Thanks! comment is now added